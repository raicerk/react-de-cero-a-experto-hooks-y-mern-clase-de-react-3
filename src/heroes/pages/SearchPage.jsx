import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useform";
import { HeroCard } from "../components/HeroCard";
import queryString from "query-string";
import { getHeroesByName } from "../helpers/getHeroesByName";

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);

  const heroes = getHeroesByName(q);

  const { searchText, onInputChange } = useForm({ searchText: q });

  const onSearchSubmit = (event) => {
    event.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input
              type="text"
              placeholder="Search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button className="btn btn-outline-primary" type="submit">
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {q === "" ? (
            <div className="alert alert-primary">Search a hero:</div>
          ) : (
            heroes.length === 0 && (
              <div className="alert alert-danger">
                No hero with <b>{q}</b>
              </div>
            )
          )}

          {heroes.map((hero) => {
            return <HeroCard key={hero.id} {...hero} />;
          })}
        </div>
      </div>
    </>
  );
};
