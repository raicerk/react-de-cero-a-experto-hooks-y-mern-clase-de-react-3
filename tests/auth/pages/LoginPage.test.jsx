import { fireEvent, render, screen } from "@testing-library/react";
import { LoginPage } from "../../../src/auth/pages/LoginPage";
import { AuthContext } from "../../../src/auth/context/AuthContext";
import { MemoryRouter } from "react-router-dom";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("Test de <LoginPage/>", () => {
  const contextValue = {
    logout: jest.fn(),
    login: jest.fn(),
  };
  test("Carga componente por defecto", () => {
    const { container } = render(
      <AuthContext.Provider value={{}}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByRole("button").innerHTML).toBe("Login");
    expect(screen.getByRole("heading", { level: 1 }).innerHTML).toBe(
      "LoginPage"
    );
  });

  test("Llamado al login", () => {
    render(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(contextValue.login).toBeCalled();
  });

  test("Hacer clic en login", () => {
    const route = "/marvel";
    jest.spyOn(Storage.prototype, "getItem").mockReturnValue(route);

    render(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </AuthContext.Provider>
    );

    fireEvent.click(screen.getByRole("button"));

    expect(mockedUseNavigate).toHaveBeenCalledWith(route, {
      replace: true,
    });
  });
});
