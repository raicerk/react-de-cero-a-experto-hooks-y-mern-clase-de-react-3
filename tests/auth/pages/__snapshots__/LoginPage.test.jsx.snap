// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test de <LoginPage/> Carga componente por defecto 1`] = `
<div>
  <div
    class="container mt-5"
  >
    <h1>
      LoginPage
    </h1>
    <hr />
    <button
      class="btn btn-primary"
    >
      Login
    </button>
  </div>
</div>
`;
